### DISCUSSION TITLE
What is the purpose of NpcSystem?

### DISCUSSION THREAD

**Discussion by kchen1023** [cmu-sei/GHOSTS#387] [Q&A] [answered]
"This is currently only used to turn the client on and off (where the client binary still runs, but does nothing). It is not used to control the client's behavior as other handlers might do."

What is the purpose of this handler (like what is an example of the use case)?

**Comment by sei-dupdyke** [comment, marked as answer]
"stop performing activities" — like browsing, creating docs, etc — basically whatever is in the timeline.

**Reply by kchen1023** [reply]
Ohhhhh so if a client currently is running a timeline for web browsing and I send a stop command via NpcSystem, does it terminate the current timeline or just put it on pause?

**Comment by sei-dupdyke** [comment]
Basically pause. 

By terminate, I assume you mean delete? That does not exist. You'd have to replace it with an empty one.

**Reply by kchen1023** [reply]
Hi Dustin,
In the out timelines directory, the web browser timeline is executed first, then the stop command is also executed. When I check the logs, there is nothing about 'NpcSystem'. Am I doing something incorrectly or misunderstanding how this handler works?
![image](https://github.com/user-attachments/assets/88cda735-fc06-4ef3-a83b-1b6e84dc24f1)

This is the web browser timeline:
```{
  "ghosts": {
    "category": "api",
    "time_lines": [
      {
        "HandlerArgs": {
          "blockflash": "true",
          "blockimages": "true",
          "blockscripts": "true",
          "blockstyles": "true",
          "incognito": "true",
          "isheadless": "false",
          "stickiness": 75,
          "stickiness-depth-max": 10000,
          "stickiness-depth-min": 5
        },
        "HandlerType": "BrowserFirefox",
        "Initial": "about:blank",
        "Loop": true,
        "TimelineEvents": [
          {
            "Command": "random",
            "CommandArgs": [
              "http://wikipedia.org",
              "http://ebay.com",
              "http://craigslist.org",
              "http://reddit.com",
              "http://instagram.com",
              "http://imdb.com",
              "http://huffingtonpost.com",
              "http://nytimes.com",
              "http://msn.com",
              "http://buzzfeed.com",
              "http://homedepot.com",
              "http://target.com",
              "http://bestbuy.com"
            ],
            "DelayAfter": "300000",
            "DelayBefore": "0"
          }
        ],
        "UtcTimeOff": "24:00:00",
        "UtcTimeOn": "00:00:00"
      }
    ]
  }
}
```

and here is the stop timeline:
```{
  "ghosts": {
    "category": "api",
    "time_lines": [
      {
        "HandlerType": "NpcSystem",
        "Initial": "",
        "Loop": false,
        "TimelineEvents": [
          {
            "Command": "Stop",
            "CommandArgs": [],
            "DelayAfter": "0",
            "DelayBefore": "0"
          }
        ],
        "UtcTimeOff": "24:00:00",
        "UtcTimeOn": "00:00:00"
      }
    ]
  }
}
```

Both of these are sent as a JSON payload.

**Reply by sei-dupdyke** [reply]
In your client, please check the logs for something like this:

`_log.Trace($"Handling NpcSystem call: {handler}");`

**Reply by kchen1023** [reply]
I checked both the app and client logs, theres no lines with NpcSystem in the client log and there's this in the app log:
![image](https://github.com/user-attachments/assets/ada6bd6f-c421-4003-8585-7d1d6e356f3d)

![image](https://github.com/user-attachments/assets/14060a02-83e3-411f-8e7a-7d23eb67c68e)

I've let it run for about 8 minutes now and the web browser timeline is still running...
